### PR TITLE
Policy for data validation worker running on Fargate

### PR DESCRIPTION
**Pull request by pyalex**
For VPC deployments similarly to the feature server policy we want to have a separate policy for data validation workers that also will be hosted on Fargate.

This policy is created when the flag `data_validation_on_fargate_enabled` is set to `true`. However, in order to use Fargate we also need "pod_execution_role" to be created. Hence `data_validation_on_fargate_enabled` should be switched on together with `fargate_enabled`.

### CODE REVIEW

File: emr_sample/outputs.tf [hunk 1]
```diff
@@ -46,6 +46,8 @@ output "roles" {
     fargate_kinesis_firehose_stream_role_name   = (var.apply_layer > 1) ? module.roles[0].fargate_kinesis_firehose_stream_role_name : ""
     fargate_eks_fargate_pod_execution_role_name = (var.apply_layer > 1) ? module.roles[0].fargate_eks_fargate_pod_execution_role_name : ""
     fargate_node_policy_name                    = (var.apply_layer > 1) ? module.roles[0].eks_fargate_node_policy_name : ""
+
+    eks_fargate_data_validation_worker_policy_name  = (var.apply_layer > 1) ? module.roles[0].eks_fargate_data_validation_worker_policy_name : ""
```
Comment (scottames) [inline emr_sample/outputs.tf:50]:
```suggestion
    fargate_data_validation_worker_policy_name  = (var.apply_layer > 1) ? module.roles[0].eks_fargate_data_validation_worker_policy_name : ""
```

File: roles/outputs.tf [hunk 2]
```diff
@@ -48,3 +48,7 @@ output "fargate_eks_fargate_pod_execution_role_name" {
 output "eks_fargate_node_policy_name" {
   value = var.fargate_enabled ? aws_iam_policy.eks_fargate_node_policy[0].name : ""
 }
+
+output "eks_fargate_data_validation_worker_policy_name" {
```
Comment (scottames) [inline roles/outputs.tf:52]:
```suggestion
output "fargate_data_validation_worker_policy_name" {
```

File: roles/variables.tf [hunk 3]
```diff
@@ -70,3 +70,14 @@ variable "fargate_enabled" {
...
+variable "data_validation_on_fargate_enabled" {
+  default     = false
+  type        = bool
+  description = <<EOT
+    Enable running data validation jobs using Fargate.
+    Otherwise they will be scheduled on EC2 machines, if data validation is enabled for cluster.
+    `fargate_enabled` should be set to true for this take effect.
+    Default: false.
+  EOT
+}
```
Comment (scottames) [inline roles/variables.tf:83]:
```suggestion
}

```

File: templates/data_validation_worker_policy.json [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+        "ecr:ListImages",
+        "ecr:DescribeImages",
+        "ecr:BatchGetImage"
+      ],
+      "Resource": [
+        "arn:aws:ecr:*:${ASSUMING_ACCOUNT_ID}:tecton/*"
+      ]
+    }
+  ]
+}
```
Comment (scottames) [inline templates/data_validation_worker_policy.json:47]:
```suggestion
}

```

**Review by scottames — CHANGES_REQUESTED** [review]
A couple small nits. Primarily, can we namespace the outputs with the other "fargate" outputs?

@yan-lgtm @kevjumba mind double checking me?

File: roles/roles.tf [hunk 5]
```diff
@@ -91,16 +101,43 @@ data "template_file" "devops_fargate_role_json" {
...
+        local.feature_server_policies,
+        local.data_validation_worker_policies
       )
     )
   }
 }
 
+# Data validation [Common : Databricks and EMR]
+data "template_file" "eks_fargate_data_validation_worker_policy" {
+  count = (var.fargate_enabled && var.data_validation_on_fargate_enabled) ? 1 : 0
```
Comment (yan-lgtm) [inline roles/roles.tf:121]: can you move `(var.fargate_enabled && var.data_validation_on_fargate_enabled)` into a variable so you can reuse?

File: roles/roles.tf [hunk 6]
```diff
@@ -91,16 +101,43 @@ data "template_file" "devops_fargate_role_json" {
...
+  vars = {
+    ACCOUNT_ID          = var.account_id
+    ASSUMING_ACCOUNT_ID = var.tecton_assuming_account_id
+    DEPLOYMENT_NAME     = var.deployment_name
+    REGION              = var.region
+  }
+}
+
+# Data validation [Common : Databricks and EMR]
+resource "aws_iam_policy" "eks_fargate_data_validation_worker_policy" {
```
Comment (yan-lgtm) [inline roles/roles.tf:133]: this is called aws_iam_policy already, can you rename this to "eks_fargate_data_validation_worker"

File: templates/data_validation_worker_policy.json [hunk 7]
```diff
@@ -0,0 +1,47 @@
+{
+  "Version": "2012-10-17",
+  "Statement": [
+    {
+      "Sid": "ReadWriteDataMetricsPolicies",
```
Comment (yan-lgtm) [inline templates/data_validation_worker_policy.json:5]: This is for dynamodb, do you need this for your role?

File: databricks_sample/infrastructure.tf [hunk 8]
```diff
@@ -110,6 +121,8 @@ module "roles" {
   tecton_assuming_account_id      = var.tecton_assuming_account_id
   elasticache_enabled             = var.elasticache_enabled
   fargate_enabled                 = var.fargate_enabled
+
+  data_validation_on_fargate_enabled = var.data_validation_on_fargate_enabled
```
Comment (yan-lgtm) [inline databricks_sample/infrastructure.tf:125]: Can you add extra spaces from line 113-123?

File: templates/data_validation_worker_policy.json (hunk 7 above)
Comment (pyalex) [reply templates/data_validation_worker_policy.json:5]: Yes, data metrics are stored in Dynamo.

Comment (pyalex) [reply templates/data_validation_worker_policy.json:5]: We read and write metrics for validation workers.

File: roles/roles.tf (hunk 5 above)
Comment (pyalex) [reply roles/roles.tf]: Added to locals

File: databricks_sample/infrastructure.tf (hunk 8 above)
Comment (pyalex) [reply databricks_sample/infrastructure.tf:124]: fomatted

File: databricks_sample/outputs.tf [hunk 9]
```diff
@@ -29,6 +29,8 @@ output "roles" {
     fargate_kinesis_firehose_stream_role_name   = module.roles.fargate_kinesis_firehose_stream_role_name
     fargate_eks_fargate_pod_execution_role_name = module.roles.fargate_eks_fargate_pod_execution_role_name
     fargate_node_policy_name                    = module.roles.eks_fargate_node_policy_name
+
+    fargate_data_validation_worker_policy_name = module.roles.fargate_data_validation_worker_policy_name
```
Comment (yan-lgtm) [inline databricks_sample/outputs.tf:33]: Can you also format here?

**Review by yan-lgtm — APPROVED** [review]

File: databricks_sample/outputs.tf (hunk 9 above)
Comment (pyalex) [reply databricks_sample/outputs.tf]: Done

**Review by yan-lgtm — APPROVED** [review]

**Review by scottames — APPROVED** [review]

**Pull request merged**